File: Sorting/FindingKthSmallestElement.py
#code
from random import randint

def QuickSelect(l, low, high, k):
    
    if low < high:
        pivot = l[high]
        pi = Partition(l, low, high, pivot)
        #print(l, pi)
        if pi == k:
            return l[k]
        elif k < pi:
            return QuickSelect(l, low, pi - 1, k)
        else:
            return QuickSelect(l, pi + 1, high, k)
    return l[k]

def Partition(l, low, high, pivot):
    i = low - 1;
    
    for j in range(low, high):
        
        if l[j] < pivot:
            i += 1
            temp = l[i]
            l[i] = l[j]
            l[j] = temp
            
    temp = l[i+1]
    l[i+1] = l[high]
    l[high] = temp
    #print("i+1 = ",i+1)
    return (i + 1);

T = int(input())

for i in range(0, T):
    n = int(input())
    l = list(map(int, input().split()))
    k = int(input())

    ele = QuickSelect(l, 0, len(l) - 1, k - 1)
    print(ele)

File: Sorting/FindingKthSmallestElement.cpp
#include<iostream>
#include<vector>
#include <climits>

using namespace std;

int partition(int arr[], int l, int r, int pivot) {
    
    int i = l - 1;

    for(int j = l; j < r; j++) {

        if(arr[j] < pivot) {
            i += 1;
            int t = arr[i];
            arr[i] = arr[j];
            arr[j] = t;
        }
    }

    int t = arr[i+1];
    arr[i+1] = arr[r];
    arr[r] = t;

    return i+1;
}

int quickSelect(int arr[], int l, int r, int k) {
    
    if(k > 0 && l < r) {

        int pivot = arr[r];

        int idx = partition(arr, l, r, pivot);

        if(idx == k )
            return arr[k];
        else if (idx < k)
            return quickSelect(arr, l, idx-1, k);
        else
            return quickSelect(arr, idx+1, r, k);
    }
    
    return arr[k];
}

int main() {

    // vector<int> arr;
    int arr[5] = {44, 23, 37, 56, 48};

    cout << quickSelect(arr, 0, 4, 2) << endl;
}